### a. 

Say you roll a die, and are given an amount in dollars equal to the number on the die

**What would you pay to play this game if you played it many times in a row?**

### b. 

Now say when you roll the die, you're allowed to either take the money, or try again

If you tried again, you have to take the second roll

**What would you pay to play this version of the game?**

### c.

Now say you can take a second roll, and if you don't like it, you get another chance to roll again

**What would you pay to play this final version?**

_____

### a.

- When we roll a six-sided die, our expected roll is 3.5
    - In the context of this game, that means we expect to win 3.50 on each roll
        - If the price to play this game was less than 3.50, we'd expect to win money over the long term
            - E.g. if the price was 3.00, we'd expect to win 50 cents each roll
                - We could play 100 games, and we'd expect to walk away with 50 bucks
        - Similarly, if the price to play was more than 3.50, we'd expect to lose money in the long run
        
- Therefore, **a fair price to pay to play this game is 3.50**

_____

### b. 

- Now that we know that the average roll will win 3.50, we know whether it's worth it to risk rolling again
    - If we roll a 1, 2, or 3, then the expected value of our next roll is higher than our current roll
        - So, in this case, we take the second roll
    - If we roll a 4, 5, or 6, we expect our next roll to be lower
        - In this case, we keep our first roll
        
- So now, we calculate the expected payoff for this game

$$
\text{Expected Payoff} = (1/2)(3.5) + (1/6)(4) + (1/6)(5) + (1/6)(6) = 4.25
$$

- Therefore, **the fair price to pay to play this game is 4.25**

- Let's confirm this result

In [3]:
import numpy as np

In [4]:
list_results = []

for roll_1 in [1,2,3,4,5,6]:
    for roll_2 in [1,2,3,4,5,6]:
        if roll_1 in [1,2,3]:
            payoff = roll_2
        else:
            payoff = roll_1
        list_results.append(payoff)
        
np.mean(list_results)

4.25

_____

### c. 

- This on requires a little bit of thinking

*First Roll = 1, 2, or 3*

- In this scenario, we'll always roll again, so the second and third rolls will follow the same logic as in part b

*First Roll = 6*

- In this scenario, we'll never roll again since we can't do any better

*First Roll = 4*

- This is where it gets tricky
    - The probability that we get a 4, 5 or 6 on my next roll is 1/2
        - So we have a 50/50 chance of staying the same or improving our roll
            - We won't roll again in this scenario, because it follows the logic of part b
        - If we get a worse roll (1, 2, or 3), we get another chance to improve with an expected roll of 3.5
        
- Therefore, if we get a 4 on my first roll and **don't roll again**, my expected payoff if 4
    - If **we do roll again**, my expected payoff is equal to:
    
$$
E(\text{Payoff} | \text{4 on first roll and we roll again}) = (4)(1/6) + (5)(1/6) + (6)(1/6) + (3.5)(1/2) = 4.25
$$

- So, it's worthwhile to roll again if we get a 4 on the first roll

*First Roll = 5*

- Our expected payoff if we roll again is the same as the one calculated above when the first roll is 4
    - In this case, however, our expected payoff for **NOT** rolling again is 5 instead of 4
    
- So, it's worthwhile to **NOT** roll again if we get a 5 on the first roll

- Instead of calculating this by hand, we can use a similar loop to the one above to calculate our expected payoff using this strategy

In [13]:
list_results = []

for roll_1 in [1,2,3,4,5,6]:
    for roll_2 in [1,2,3,4,5,6]:
        for roll_3 in [1,2,3,4,5,6]:
            if roll_1 in [1,2,3] and roll_2 in [1,2,3]:
                payoff = roll_3
            elif roll_1 in [1,2,3] and roll_2 in [4,5,6]:
                payoff = roll_2
            elif roll_1 in [5,6]:
                payoff = roll_1
            else:
                if roll_2 in [1,2,3]:
                    payoff = roll_3
                else:
                    payoff = roll_2
            list_results.append(payoff)           
        
np.mean(list_results)

4.666666666666667

- So, following this strategy, the **fair price to play is 4.67**